modules/mesh-layers/src/utils/matrix.js

```diff
@@ -11,118 +17,144 @@ function updateRotationMatrix(targetMatrix, roll, pitch, yaw) {
+function calculateModelMatrices(layer, attribute) {
+  const {data, getOrientation, getScale, getTranslation, getTransformMatrix} = layer.props;
+
+  const constantMatrix = Array.isArray(getTransformMatrix);
+  const constantScale = Array.isArray(getScale);
+  const constantOrientation = Array.isArray(getOrientation);
+  const constantTranslation = Array.isArray(getTranslation);
+
+  const hasMatrix = getTransformMatrix && (constantMatrix || Boolean(getTransformMatrix(data[0])));
+
+  if (hasMatrix) {
+    attribute.constant = constantMatrix;
+  } else {
+    attribute.constant = constantOrientation && constantScale && constantTranslation;
+  }
 
   const instanceModelMatrixData = attribute.value;
 
-  const rotationMatrix = new Float32Array(16);
-  const modelMatrix = new Matrix4();
-
-  let i = 0;
-  for (const object of data) {
-    const matrix =
-      getMatrix(object) ||
-      makeMatrix({
-        object,
-        getYaw,
-        getPitch,
-        getRoll,
-        getScale,
-        getTranslation,
-        rotationMatrix,
-        modelMatrix
-      });
-
-    instanceModelMatrixData[i++] = matrix[0];
-    instanceModelMatrixData[i++] = matrix[1];
-    instanceModelMatrixData[i++] = matrix[2];
-    instanceModelMatrixData[i++] = matrix[3];
-    instanceModelMatrixData[i++] = matrix[4];
-    instanceModelMatrixData[i++] = matrix[5];
-    instanceModelMatrixData[i++] = matrix[6];
-    instanceModelMatrixData[i++] = matrix[7];
-    instanceModelMatrixData[i++] = matrix[8];
-    instanceModelMatrixData[i++] = matrix[9];
-    instanceModelMatrixData[i++] = matrix[10];
-    instanceModelMatrixData[i++] = matrix[11];
-    instanceModelMatrixData[i++] = matrix[12];
-    instanceModelMatrixData[i++] = matrix[13];
-    instanceModelMatrixData[i++] = matrix[14];
-    instanceModelMatrixData[i++] = matrix[15];
+  if (attribute.constant) {
+    let matrix;
+
+    if (hasMatrix) {
+      modelMatrix.set(getTransformMatrix);
+      modelTranslation[0] = modelMatrix[12];
+      modelTranslation[1] = modelMatrix[13];
+      modelTranslation[2] = modelMatrix[14];
+      matrix = getMat3FromMat4(modelMatrix);
+    } else {
+      matrix = linearTransform;
+
+      const orientation = getOrientation;
+      const scale = getScale;
+
+      calculateTransformMatrix(matrix, orientation, scale);
+      modelTranslation.set(getTranslation);
+    }
+
+    const valueMatrix = new Float32Array(matrix);
+    const valueTranslation = new Float32Array(modelTranslation);
+    const shaderAttributes = attribute.userData.shaderAttributes;
+    shaderAttributes.instanceModelMatrix__LOCATION_0.value = valueMatrix.subarray(0, 3);
+    shaderAttributes.instanceModelMatrix__LOCATION_1.value = valueMatrix.subarray(3, 6);
+    shaderAttributes.instanceModelMatrix__LOCATION_2.value = valueMatrix.subarray(6, 9);
+    shaderAttributes.instanceTranslation.value = valueTranslation;
+  } else {
+    let i = 0;
+    for (const object of data) {
```

Comment:
We need to handle binary data here - See https://github.com/uber/deck.gl/blob/master/modules/layers/src/scatterplot-layer/scatterplot-layer.js#L171